File: typescript/packages/core/src/process/supervisor.ts
import {
  DEFAULT_PROCESS_PERMISSIONS,
  type ProcessPermissions,
  type ProcessScope,
} from './config.ts'
import type { PathSpec } from '../types.ts'
import { ProcessHandle } from './handle.ts'
import type { ProcessRunner, ProcessView } from './types.ts'

/** Workspace-owned live runners; exited handles retain their own results. */
export class ProcessSupervisor {
  private nextPid = 1
  private readonly runners = new Map<number, { generation: number; handle: ProcessHandle }>()
  private readonly generations = new Map<string, number>()
  private stopped = false

  /** Host-only: the supplied runner must admit commands before effects. */
  start(init: {
    sessionId: string
    command: string
    cwd: PathSpec
    run: ProcessRunner
    cancel: () => void
    parentPid?: number | null
  }): ProcessHandle {
    if (this.stopped) throw new Error('process supervisor is stopped')
    const parent = init.parentPid == null ? undefined : this.runners.get(init.parentPid)
    if (
      init.parentPid != null &&
      (parent === undefined || parent.handle.info.cancellationRequested)
    )
      throw new Error('parent process is no longer accepting children')
    const pid = this.nextPid++
    const handle = new ProcessHandle(
      {
        pid,
        parentPid: init.parentPid ?? null,
        groupId:
          (init.parentPid == null
            ? undefined
            : this.runners.get(init.parentPid)?.handle.info.groupId) ?? pid,
        sessionId: init.sessionId,
        command: init.command,
        cwd: init.cwd,
        startedAt: Date.now() / 1000,
        state: 'running',
        cancellationRequested: false,
        exitCode: null,
        failure: null,
      },
      init.run,
      () => {
        for (const child of this.live())
          if (
            child.info.parentPid === pid ||
            (child.info.groupId === pid && child.info.pid !== pid)
          )
            child.terminate()
        init.cancel()
      },
      (id) => {
        this.runners.delete(id)
      },
    )
    this.runners.set(pid, { generation: this.generations.get(init.sessionId) ?? 0, handle })
    return handle
  }

  view(
    sessionId: string,
    permissions: () => ProcessPermissions = () => DEFAULT_PROCESS_PERMISSIONS,
  ): ProcessView {
    const generation = this.generations.get(sessionId) ?? 0
    const valid = () => (this.generations.get(sessionId) ?? 0) === generation
    const allowed = (scope: ProcessScope, handle: ProcessHandle) =>
      scope === 'workspace' || (scope === 'session' && handle.info.sessionId === sessionId)
    const visible = (handle: ProcessHandle) => {
      const entry = this.runners.get(handle.info.pid)
      if (
        handle.info.sessionId === sessionId &&
        entry !== undefined &&
        entry.generation !== generation
      )
        return null
      const grants = permissions()
      if (!valid() || !allowed(grants.metadata, handle)) return null
      return allowed(grants.details, handle)
        ? handle.info
        : Object.freeze({ ...handle.info, command: null, cwd: null, failure: null })
    }
    return Object.freeze({
      list: () =>
        Object.freeze(
          [...this.runners.values()].flatMap(({ handle }) => {
            const info = visible(handle)
            return info === null ? [] : [info]
          }),
        ),
      get: (pid: number) => {
        const entry = this.runners.get(pid)
        return entry === undefined ? null : visible(entry.handle)
      },
      checkSpawn: () => {
        if (!valid() || !permissions().spawn)
          throw Object.assign(new Error('process spawn is not permitted'), { code: 'EACCES' })
      },
      terminate: (pid: number) => {
        const entry = this.runners.get(pid)
        return (
          entry !== undefined &&
          visible(entry.handle) !== null &&
          allowed(permissions().control, entry.handle) &&
          entry.handle.terminate()
        )
      },
      wait: async (pid: number) => {
        const entry = this.runners.get(pid)
        if (entry === undefined || visible(entry.handle) === null) return null
        await entry.handle.join()
        return visible(entry.handle)
      },
    })
  }

  revokeSession(sessionId: string): void {
    this.generations.set(sessionId, (this.generations.get(sessionId) ?? 0) + 1)
  }

  /** Host-only inventory, including disowned and stopping runners. */
  live(): readonly ProcessHandle[] {
    return [...this.runners.values()].map((entry) => entry.handle)
  }

  /** Join managed runners before releasing their workspace resources. */
  async drain(): Promise<void> {
    await Promise.all(this.live().map((process) => process.join()))
  }

  /** Close admission and request cancellation; stopping is not completion. */
  stop(): void {
    this.stopped = true
    const errors: unknown[] = []
    for (const process of this.live()) {
      try {
        process.terminate()
      } catch (error) {
        errors.push(error)
      }
    }
    if (errors.length > 0) throw new AggregateError(errors, 'process cancellation failed')
  }
}

File: python/mirage/process/supervisor.py
import asyncio
import time
from collections.abc import Callable
from dataclasses import replace
from threading import RLock

from mirage.process.config import ProcessPermissions, ProcessScope
from mirage.process.handle import ProcessHandle
from mirage.process.types import ProcessInfo, ProcessRunner, ProcessView
from mirage.types import PathSpec


class ProcessSupervisor:
    """Own live runners independently of shell job-table membership.

    IDs are workspace-local and never reused. Exited runners leave this
    live registry; holders of their handles retain the final result.
    """

    def __init__(self) -> None:
        self._next_pid = 1
        self._live: dict[int, tuple[int, ProcessHandle]] = {}
        self._generations: dict[str, int] = {}
        self._lock = RLock()
        self._stopped = False

    def start(self,
              *,
              session_id: str,
              command: str,
              cwd: PathSpec,
              run: ProcessRunner,
              parent_pid: int | None = None) -> ProcessHandle:
        """Track a runner; its caller owns admission before command effects.

        Args:
            session_id (str): owner session, never a profile label.
            command (str): display text, never parsed by this layer.
            cwd (PathSpec): working directory at launch.
            run (ProcessRunner): asynchronous execution and cleanup.
        """
        with self._lock:
            if self._stopped:
                raise RuntimeError("process supervisor is stopped")
            pid = self._next_pid
            self._next_pid += 1
            parent = self._live.get(
                parent_pid) if parent_pid is not None else None
            if parent_pid is not None and (
                    parent is None or parent[1].info.cancellation_requested):
                raise RuntimeError(
                    "parent process is no longer accepting children")
            group_id = parent[1].info.group_id if parent is not None else pid
            handle = ProcessHandle(
                ProcessInfo(pid,
                            session_id,
                            command,
                            cwd,
                            time.time(),
                            parent_pid=parent_pid,
                            group_id=group_id), run, self._retire,
                lambda: self.terminate_children(pid))
            self._live[pid] = (self._generations.get(session_id, 0), handle)
            return handle

    def terminate_children(self, pid: int) -> None:
        """Request cancellation of every live runner under ``pid``.

        A runner is under ``pid`` when ``pid`` is its parent or started
        its execution group, so grandchildren are reached after an
        intermediate runner has exited.

        Args:
            pid (int): the parent or group leader.
        """
        for child in self.live():
            if child.info.parent_pid == pid or (child.info.group_id == pid
                                                and child.info.pid != pid):
                child.terminate()

    def _retire(self, pid: int) -> None:
        with self._lock:
            self._live.pop(pid, None)

    def view(
        self,
        session_id: str,
        permissions: Callable[[], ProcessPermissions] = ProcessPermissions
    ) -> ProcessView:
        with self._lock:
            generation = self._generations.get(session_id, 0)

        def valid() -> bool:
            return self._generations.get(session_id, 0) == generation

        def allowed(scope: ProcessScope, handle: ProcessHandle) -> bool:
            return scope == "workspace" or (
                scope == "session" and handle.info.session_id == session_id)

        def visible(handle: ProcessHandle) -> ProcessInfo | None:
            entry = self._live.get(handle.info.pid)
            if (handle.info.session_id == session_id and entry is not None
                    and entry[0] != generation):
                return None
            grants = permissions()
            if not valid() or not allowed(grants.metadata, handle):
                return None
            info = handle.info
            return info if allowed(grants.details, handle) else replace(
                info, command=None, cwd=None, failure=None)

        def list_visible() -> tuple[ProcessInfo, ...]:
            with self._lock:
                return tuple(info for _, handle in self._live.values()
                             if (info := visible(handle)) is not None)

        def get_visible(pid: int) -> ProcessInfo | None:
            with self._lock:
                entry = self._live.get(pid)
                return visible(entry[1]) if entry is not None else None

        def check_spawn() -> None:
            if not valid() or not permissions().spawn:
                raise PermissionError("process spawn is not permitted")

        def terminate(pid: int) -> bool:
            with self._lock:
                entry = self._live.get(pid)
                if entry is None or visible(entry[1]) is None or not allowed(
                        permissions().control, entry[1]):
                    return False
                return entry[1].terminate()

        async def wait(pid: int) -> ProcessInfo | None:
            with self._lock:
                entry = self._live.get(pid)
                if entry is None or visible(entry[1]) is None:
                    return None
                handle = entry[1]
            await handle.join()
            return visible(handle)

        return ProcessView(list=list_visible,
                           get=get_visible,
                           check_spawn=check_spawn,
                           terminate=terminate,
                           wait=wait)

    def revoke_session(self, session_id: str) -> None:
        """Revoke the session's views and cancel its runners.

        A closed session's ID can be reused and a replaced profile grants
        a new view, so neither may keep a door, or a runner admitted
        under the old grants.

        Args:
            session_id (str): session being closed or re-profiled.
        """
        with self._lock:
            self._generations[session_id] = self._generations.get(
                session_id, 0) + 1
        for process in self.live():
            if process.info.session_id == session_id:
                process.terminate()

    def live(self) -> tuple[ProcessHandle, ...]:
        """Host-only inventory, including disowned and stopping runners."""
        with self._lock:
            return tuple(handle for _, handle in self._live.values())

    async def drain(self) -> None:
        """Join managed runners before releasing their workspace resources."""
        await asyncio.gather(*(process.join() for process in self.live()))

    def stop(self) -> None:
        """Close admission and request cancellation of all remaining runners.

        This does not join: uncooperative runners remain visible as stopping.
        """
        with self._lock:
            self._stopped = True
        errors: list[Exception] = []
        for process in self.live():
            try:
                process.terminate()
            except Exception as exc:
                errors.append(exc)
        if errors:
            raise ExceptionGroup("process cancellation failed", errors)
